vllm/model_executor/layers/quantization/fp8fnuz.py

```diff
@@ -0,0 +1,262 @@
+                                 layer=layer,
+                                 output_partition_sizes=output_partition_sizes,
+                                 **extra_weight_attrs)
+
+        self._create_scale_param(scale_name="activation_scaling_factor",
+                                 layer=layer,
+                                 output_partition_sizes=output_partition_sizes,
+                                 **extra_weight_attrs)
+
+        self._create_scale_param(scale_name="output_scaling_factor",
+                                 layer=layer,
+                                 output_partition_sizes=output_partition_sizes,
+                                 **extra_weight_attrs)
+
+    def process_weights_after_loading(self, layer: Module) -> None:
+        if (not hasattr(layer, "process_after_load")
+                or not layer.process_after_load):
+            return
+
+        layer.activation_scaling_factor = Parameter(
+            layer.activation_scaling_factor.max(), requires_grad=False)
+        layer.output_scaling_factor = Parameter(
+            layer.output_scaling_factor.reciprocal().max(),
+            requires_grad=False)
+
+        max_w_scale = layer.weight_scaling_factor.max()
+        if len(layer.logical_widths) > 1:
+            start = 0
+            for idx, logical_width in enumerate(layer.logical_widths):
+                end = start + logical_width
+                weight_dq = _per_tensor_dequantize(
+                    layer.weight[start:end, :],
+                    layer.weight_scaling_factor[idx])
+
+                layer.weight[start:end, :] = _per_tensor_quantize(
+                    weight_dq, max_w_scale)
+                start = end
+        layer.weight_scaling_factor = Parameter(max_w_scale,
+                                                requires_grad=False)
+
+        # WEIGHT
+        weight = layer.weight
+        layer.weight = Parameter(weight, requires_grad=False)
+
+    def scales_shard_indexer(
+            self, param: torch.Tensor, loaded_weight: torch.Tensor,
+            shard_id: Union[str, int]) -> Tuple[torch.Tensor, torch.Tensor]:
+        qkv_idxs = {"q": 0, "k": 1, "v": 2}
+
+        if isinstance(shard_id, int):
+            pass
+        elif isinstance(shard_id, str):
+            if shard_id not in qkv_idxs:
+                raise ValueError(f"Unknown shard_id: {shard_id}")
+            shard_id = qkv_idxs[shard_id]
+        else:
+            ValueError(f"Shard id must be int or str but got {type(shard_id)}")
+
+        # To handle the scalar loaded tensor
+        if loaded_weight.numel() == 1 and len(loaded_weight.shape) != 0:
+            loaded_weight = torch.scalar_tensor(loaded_weight[0])
+
+        return param[shard_id], loaded_weight
+
+    def apply(
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        weight: torch.Tensor = layer.weight
+
```

Comment:
This should all be pushed into `apply_fp8_linear` such that the other integrations can run on the AMD backend!